src/hotspot/share/jfr/recorder/checkpoint/types/traceid/jfrTraceIdKlassQueue.cpp

```diff
@@ -75,13 +75,14 @@ static size_t element_size(bool compressed) {
   return compressed ? NARROW_ELEMENT_SIZE : ELEMENT_SIZE;
 }
 
-static bool can_compress_element(traceid id) {
-  return Metaspace::using_class_space() && id < uncompressed_threshold;
+static bool can_compress_element(const Klass* klass) {
+  return Metaspace::is_in_class_space(klass) &&
```

Comment:
```suggestion
  return (Metaspace::is_in_class_space(klass) || Metaspace::is_in_shared_metaspace(klass)) &&
```

Reply:
Is this right?  If UseCompressedClassPointers is off, then the shared metaspace isn't in compressed space?

Reply:
With UseCompressedClassPointers off, I think Metaspace::is_in_shared_metaspace() would still return true but I don't think he compression base is the bottom of the CDS archive.  I asked Markus to have a look.

Reply:
The code supports the JfrTraceID load barrier that enqueues tagged Klass*. It selects a more compact representation (a single word, instead of two words), if a Klass* can be compressed (i.e. there exists a compress class scheme in place (CompressedKlassPointers::encode(const_cast<Klass*>(klass)); AND the traceid (u8) value is low enough to be represented by only 4 bytes).

struct JfrEpochQueueKlassElement {
  traceid id;
  const Klass* klass;
};

struct JfrEpochQueueNarrowKlassElement {
  u4 id;
  narrowKlass compressed_klass;
};

Reply:
// Return TRUE only if UseCompressedClassPointers is True.
  static bool using_class_space() {
    return NOT_LP64(false) LP64_ONLY(UseCompressedClassPointers);
  }
  
  I see now that was wrong for 32-bit.

Reply:
In summary, we are agnostic about which space the Klass* is located in; we only care if a valid means exists to perform an encode() and decode() operation to compress the Klass* (for 64-bit to be clear).

Reply:
Its very rare, if at all, that an abstract or an interface would be tagged in JFR. Tags are for concrete implementations, mostly InstanceKlass*.

Reply:
I now read the JIRA issue. JFR do process loads of  java/lang/invoke/LambdaForm$MH and derivatives.

Reply:
It's fine to have each Klass* report whether it can be compressed. If not, it will be represented using the non-compressed version, which will be a bit more bloated, but no problems.

Reply:
narrowKlass is the result of encoding Klass* with CompressedKlassPointers::encode() which is relative to the compressed base, so if UseCompressedClassPointers is false then the encoding to narrowKlass from some other (CDS?) base isn't valid.
using_class_space() above doesn't look wrong for 32 bits.  It should return false.

With this patch, interface and abstract classes cannot be encoded and decoded to yield a valid compressed narrowKlass, since they're now allocated in the non-class metaspace.  Yes, this now a function of which space the Klass resides in.  We do crash for these classes in JFR without this change.

Reply:
I mean its wrong from JFRs perspective not to handle 32-bit outside of the call to using_class_space(), because that call will always be false for 32-bit, although the Klass* will still fit in 4-bytes.

Reply:
Looks ok from a JFR perspective, Coleen.

Reply:
Thank you Markus.

Reply:
Many thanks for explaining this, @mgronlun . 

@coleenp I think your code is correct insofar as it does not break anything. Your version misses out on compression opportunities for Klass in CDS, but that may not be such a big deal. You were right though that my variant was incorrect, too.